File: Sort/C/mergesort.c
#include "main.h"
/**
 * merge - merges subarrays 
 *
 * Return: void
 */
void merge(int arr[], int b, int m, int e)
{
	int i, j, k;
	
	int n1 = m - b + 1;
	int n2 = e - m;

	int L[n1], R[n2];

	for (i = 0; i < n1; i++)
	{
		L[i] = arr[b + i];
	}
	
	for ( j = 0; j < n2; j++)
	{
		R[j] = arr[m + 1 + j];
	}
	
	k = b;
	i = 0;
	j = 0;
	while (i < n1 && j < n2)
	{
		if (L[i] <= R[j])
		{
			arr[k] = L[i];
			i++;
		}
		else
		{
			arr[k] = R[j];
			j++;
		}
		k++;
	}
	
	while (i < n1)
	{
		arr[k] = L[i];
		i++;
		k++;
	}
	
	while (j <n2)
	{
		arr[k] = R[j];
		j++;
		k++;
	}
	
}
/**
 * mergesort - sortingtechnique
 *
 * Return: void
 */
void mergesort(int arr[], int beg, int end)
{
	int mid;

	if (beg < end)
	{
		mid = beg + (end - beg) / 2;
		mergesort(arr, beg, mid);
		mergesort(arr, mid + 1, end);
		merge(arr, beg, mid, end);
	}
}

File: Sort/Java/MergeSort.java
public class MergeSort {
public static class MSort {
	public void merge(int arr[], int beg, int mid, int end)
	{
		int n1 = mid - beg + 1;
		int n2 = end - mid;
		int L[] = new int[n1];
		int R[] = new int[n2];

		for (int i = 0; i < n1; ++i)
			L[i] = arr[beg + i];
		for (int j = 0; j < n2; ++j)
			R[j] = arr[mid + 1 + j];

		int i = 0, j = 0;
		int k = beg;
		while (i < n1 && j < n2) {
			if (L[i] <= R[j]) {
				arr[k] = L[i];
				i++;
			}
			else {
				arr[k] = R[j];
				j++;
			}
			k++;
		}
		while (i < n1) {
			arr[k] = L[i];
			i++;
			k++;
		}
		while (j < n2) {
			arr[k] = R[j];
			j++;
			k++;
		}
	}
	void sort(int arr[], int beg, int end)
	{
		if (beg < end) {
			int mid = beg + (end - beg) / 2;

			sort(arr, beg, mid);
			sort(arr, mid + 1, end);
			merge(arr, beg, mid, end);
		}
	}
}
    public static void printArray(int arr[])
    {
        int n = arr.length;
        for (int i = 0; i < n; ++i)
            System.out.print(arr[i] + " ");
        System.out.println();
    }
    public static void main(String args[])
    {
        int arr[] = { 12, 11, 13, 5, 6, 7 };

        System.out.println("Given Array");
        printArray(arr);

        MSort ob = new MSort();
        ob.sort(arr, 0, arr.length - 1);

        System.out.println("\nSorted array");
        printArray(arr);
    }
}
